atc/gc/build_log_collector.go

```diff
@@ -156,26 +148,50 @@ func (br *buildLogCollector) reapLogsOfJob(pipeline db.Pipeline,
 			}
 		}
 
-		// If Builds is 0, then all builds are retained, so we don't need to
-		// check MinSuccessBuilds at all.
-		if logRetention.Builds > 0 {
-			if logRetention.MinimumSucceededBuilds > 0 && build.Status() == db.BuildStatusSucceeded {
-				if retainedSucceededBuilds < logRetention.MinimumSucceededBuilds {
-					retainedBuilds++
-					retainedSucceededBuilds++
-					firstLoggedBuildID = build.ID()
-					continue
-				}
+		// If the minimum is set but not satisfied it should not be reaped
+		if logRetention.Builds != 0 && logRetention.MinimumSucceededBuilds != 0 {
```

Comment:
I feel like this logic can be simplified a little. All we really care about is that if any of the enabled conditions are *not* met, then we *don't* reap the build. So, what if we wrote it like this pseudocode:

```
if logRetention.Builds != 0 {
  if logRetention.MinimumSucceededBuilds != 0 {
    if build succeeded && haven't retained enough succeeded builds {
      retain(build)
      continue
    }
  }
  if haven't retained enough builds {
    retain(build)
  }
}

if logRetention.Days != 0 {
  if build has not expired {
    retain(build)
    continue
  }
}

// at this point, we haven't met all of the enabled conditions, so here we can reap
buildIDsToDelete = append(buildIDsToDelete, build.ID())
```

<details>

<summary>Some actual code to this effect</summary>

```diff
diff --git a/atc/gc/build_log_collector.go b/atc/gc/build_log_collector.go
index f0ac616a5..91b3e8a6d 100644
--- a/atc/gc/build_log_collector.go
+++ b/atc/gc/build_log_collector.go
@@ -133,6 +133,18 @@ func (br *buildLogCollector) reapLogsOfJob(pipeline db.Pipeline,
 	retainedBuilds := 0
 	retainedSucceededBuilds := 0
 	firstLoggedBuildID := 0
+
+	retain := func(build db.Build) {
+		firstLoggedBuildID = build.ID()
+		retainedBuilds++
+		if build.Status() == db.BuildStatusSucceeded {
+			retainedSucceededBuilds++
+		} else if logRetention.Builds != 0 {
+			// only need to track if there's a limit on number of builds
+			toRetainNonSucceededBuildIDs = append(toRetainNonSucceededBuildIDs, build.ID())
+		}
+	}
+
 	for _, build := range buildsToConsiderDeleting {
 		// Running build should not be reaped.
 		if build.IsRunning() {
@@ -148,51 +160,29 @@ func (br *buildLogCollector) reapLogsOfJob(pipeline db.Pipeline,
 			}
 		}
 
-		// If the minimum is set but not satisfied it should not be reaped
-		if logRetention.Builds != 0 && logRetention.MinimumSucceededBuilds != 0 {
-			if build.Status() == db.BuildStatusSucceeded && retainedSucceededBuilds < logRetention.MinimumSucceededBuilds {
-				retainedBuilds++
-				retainedSucceededBuilds++
-				firstLoggedBuildID = build.ID()
-				continue
+		if logRetention.Builds != 0 {
+			if logRetention.MinimumSucceededBuilds != 0 {
+				if build.Status() == db.BuildStatusSucceeded && retainedSucceededBuilds < logRetention.MinimumSucceededBuilds {
+					retain(build)
+					continue
+				}
 			}
-		}
-
-		maxBuildsRetained := retainedBuilds >= logRetention.Builds
 
-		// If only builds is set
-		if logRetention.Days == 0 {
-			if maxBuildsRetained {
-				logger.Debug("should-reap-due-to-builds", build.LagerData())
-				buildIDsToDelete = append(buildIDsToDelete, build.ID())
-			} else {
-				retainedBuilds++
-				toRetainNonSucceededBuildIDs = append(toRetainNonSucceededBuildIDs, build.ID())
-				firstLoggedBuildID = build.ID()
+			if retainedBuilds < logRetention.Builds {
+				retain(build)
+				continue
 			}
-			continue
 		}
 
-		buildHasExpired := !build.EndTime().IsZero() && build.EndTime().AddDate(0, 0, logRetention.Days).Before(time.Now())
-
-		// If only Days is set
-		if logRetention.Builds == 0 {
-			if buildHasExpired {
-				logger.Debug("should-reap-due-to-days", build.LagerData())
-				buildIDsToDelete = append(buildIDsToDelete, build.ID())
+		if logRetention.Days != 0 {
+			buildHasExpired := !build.EndTime().IsZero() && build.EndTime().AddDate(0, 0, logRetention.Days).Before(time.Now())
+			if !buildHasExpired {
+				retain(build)
+				continue
 			}
-			continue
 		}
 
-		// If Builds and Days are set
-		if maxBuildsRetained && buildHasExpired {
-			logger.Debug("should-reap-due-to-days-and-builds", build.LagerData())
-			buildIDsToDelete = append(buildIDsToDelete, build.ID())
-		} else {
-			retainedBuilds++
-			toRetainNonSucceededBuildIDs = append(toRetainNonSucceededBuildIDs, build.ID())
-			firstLoggedBuildID = build.ID()
-		}
+		buildIDsToDelete = append(buildIDsToDelete, build.ID())
 	}
 
 	logger.Debug("after-second-round-filter", lager.Data{
@@ -207,7 +197,7 @@ func (br *buildLogCollector) reapLogsOfJob(pipeline db.Pipeline,
 
 	// If this happens, firstLoggedBuildID must points to a success build, thus
 	// no need to update firstLoggedBuildID.
-	if retainedBuilds > logRetention.Builds {
+	if logRetention.Builds != 0 && retainedBuilds > logRetention.Builds {
 		logger.Debug("more-builds-to-retain", lager.Data{
 			"retained_builds": retainedBuilds,
 		})
```
</details>

IMO it's a bit easier to understand since we can consider the rules around builds and days independently, rather than considering each permutation (Builds+Days, Builds, Days).

WDYT?